File: quiz-engine/quiz-engine-nodejs/src/cli/Formatter.ts
import Table from 'cli-table3';
import chalk from 'chalk';

export interface ReviewItem {
  questionText: string;
  shuffledOptions: string[];
  correctShuffledIndex: number;
  userAnswerLetter: string;
  isCorrect: boolean;
  explanation?: string;
}

export class Formatter {
  /**
   * Render a table from an array of objects. All objects should have the same keys.
   */
  static table(data: Record<string, unknown>[]): string {
    if (data.length === 0) return chalk.yellow('No data to display.');

    const headers = Object.keys(data[0]);
    const table = new Table({
      head: headers.map((h) => chalk.cyan(h)),
      style: { head: [], border: [] },
    });

    for (const row of data) {
      table.push(headers.map((h) => String(row[h] ?? '')));
    }

    return table.toString();
  }

  /**
   * Render a bordered box with a title and content.
   */
  static box(title: string, content: string): string {
    const lines = content.split('\n');
    const maxWidth = Math.max(title.length + 4, ...lines.map((l) => l.length + 4), 40);
    const border = '─'.repeat(maxWidth - 2);

    const header = `┌${border}┐`;
    const footer = `└${border}┘`;
    const titleLine = `│ ${chalk.bold(title)}${' '.repeat(maxWidth - title.length - 3)}│`;
    const separator = `├${border}┤`;
    const contentLines = lines.map(
      (l) => `│ ${l}${' '.repeat(Math.max(0, maxWidth - l.length - 3))}│`,
    );

    return [header, titleLine, separator, ...contentLines, footer].join('\n');
  }

  static reviewSection(items: ReviewItem[]): string {
    const letters = ['A', 'B', 'C', 'D', 'E'];
    const lines: string[] = [];
    lines.push(chalk.bold('\n📝 Answer Review'));
    lines.push('─'.repeat(60));

    for (let i = 0; i < items.length; i++) {
      const item = items[i];
      const marker = item.isCorrect ? chalk.green('✓') : chalk.red('✗');
      const firstLine = item.questionText.split('\n')[0];
      lines.push(`${marker} Q${i + 1}: ${firstLine}`);

      if (!item.isCorrect) {
        const correctLetter = letters[item.correctShuffledIndex];
        const correctText = item.shuffledOptions[item.correctShuffledIndex];
        lines.push(
          chalk.dim(`   Your answer: ${item.userAnswerLetter} | Correct: ${correctLetter}) ${correctText}`),
        );
        if (item.explanation) {
          lines.push(chalk.yellow(`   💡 ${item.explanation}`));
        }
      }
    }
    lines.push('');
    return lines.join('\n');
  }

  /**
   * Format a quiz result summary.
   */
  static quizResult(
    numCorrect: number,
    numQuestions: number,
    percentage: number,
    grade: string,
    passed: boolean,
    duration: string,
  ): string {
    const lines: string[] = [];
    lines.push(chalk.bold('\n📊 Quiz Results'));
    lines.push('─'.repeat(40));
    lines.push(`Score:    ${chalk.bold(`${numCorrect} / ${numQuestions}`)}`);
    lines.push(`Percent:  ${chalk.bold(`${percentage}%`)}`);
    lines.push(`Grade:    ${chalk.bold(grade)}`);
    lines.push(`Status:   ${passed ? chalk.green('✅ PASSED') : chalk.red('❌ FAILED')}`);
    lines.push(`Duration: ${duration}`);
    lines.push('─'.repeat(40));
    return lines.join('\n');
  }

  /**
   * Format a single question for display.
   */
  static question(
    questionNumber: number,
    totalQuestions: number,
    questionText: string,
    options: string[],
  ): string {
    const letters = ['A', 'B', 'C', 'D', 'E'];
    const lines: string[] = [];
    lines.push(chalk.bold(`\nQuestion ${questionNumber} of ${totalQuestions}`));
    lines.push('─'.repeat(60));
    lines.push(chalk.white(questionText));
    lines.push('');
    options.forEach((opt, idx) => {
      lines.push(`  ${chalk.cyan(letters[idx])}) ${opt}`);
    });
    lines.push('');
    return lines.join('\n');
  }

  /**
   * Format answer feedback.
   */
  static answerFeedback(
    isCorrect: boolean,
    correctLetter: string,
    correctText: string,
    explanation?: string,
  ): string {
    const lines: string[] = [];
    if (isCorrect) {
      lines.push(chalk.green('✅ Correct!'));
    } else {
      lines.push(chalk.red(`❌ Incorrect. Correct answer: ${correctLetter}) ${correctText}`));
    }
    if (explanation) {
      lines.push(chalk.dim(`💡 ${explanation}`));
    }
    lines.push('');
    return lines.join('\n');
  }
}

File: quiz-engine/quiz-engine-rust/src/cli/formatter.rs
/// Format a fixed-width table with box-drawing characters.
pub fn print_table(headers: &[&str], rows: &[Vec<String>]) {
    if rows.is_empty() {
        println!("(no records)");
        return;
    }

    let col_count = headers.len();
    let mut widths: Vec<usize> = headers.iter().map(|h| h.len()).collect();

    for row in rows {
        for (i, cell) in row.iter().enumerate() {
            if i < col_count {
                widths[i] = widths[i].max(cell.len());
            }
        }
    }

    let separator: String = widths
        .iter()
        .map(|w| "─".repeat(w + 2))
        .collect::<Vec<_>>()
        .join("┼");

    let top: String = widths
        .iter()
        .map(|w| "─".repeat(w + 2))
        .collect::<Vec<_>>()
        .join("┬");

    let bottom: String = widths
        .iter()
        .map(|w| "─".repeat(w + 2))
        .collect::<Vec<_>>()
        .join("┴");

    println!("┌{top}┐");

    let header_row: String = headers
        .iter()
        .zip(widths.iter())
        .map(|(h, w)| format!(" {h:<w$} "))
        .collect::<Vec<_>>()
        .join("│");
    println!("│{header_row}│");

    println!("├{separator}┤");

    for row in rows {
        let cells: String = (0..col_count)
            .map(|i| {
                let cell = row.get(i).map(|s| s.as_str()).unwrap_or("");
                format!(" {cell:<width$} ", width = widths[i])
            })
            .collect::<Vec<_>>()
            .join("│");
        println!("│{cells}│");
    }

    println!("└{bottom}┘");
}

/// Print a labeled result box.
pub fn print_result_box(title: &str, lines: &[String]) {
    let max_width = lines
        .iter()
        .map(|l| l.len())
        .max()
        .unwrap_or(0)
        .max(title.len());

    let border = "═".repeat(max_width + 2);
    println!("╔{border}╗");
    println!("║ {title:<max_width$} ║");
    println!("╠{border}╣");
    for line in lines {
        println!("║ {line:<max_width$} ║");
    }
    println!("╚{border}╝");
}

/// Print a simple divider line.
pub fn print_divider(width: usize) {
    println!("{}", "─".repeat(width));
}

/// Print the end-of-quiz answer review.
/// Each entry: (question_number, question_text, user_letter, correct_letter, correct_answer_text, is_correct, explanation)
pub fn print_review(items: &[(usize, &str, &str, &str, &str, bool, Option<&str>)]) {
    if items.is_empty() {
        return;
    }

    println!();
    let border = "═".repeat(60);
    println!("╔{border}╗");
    println!("║ {:<58} ║", "Answer Review");
    println!("╚{border}╝");

    for &(num, question, user_letter, correct_letter, correct_text, is_correct, explanation) in items {
        println!();
        let marker = if is_correct { "✓" } else { "✗" };
        println!("  {marker} Q{num}: {question}");
        if is_correct {
            println!("    Your answer: {user_letter})  ✓");
        } else {
            println!("    Your answer:    {user_letter})");
            println!("    Correct answer: {correct_letter}) {correct_text}");
        }
        if let Some(expl) = explanation {
            println!("    Explanation: {expl}");
        }
    }
    println!();
}
